File: 13-eval-rpn/solution.js
// Evaluate Reverse Polish Notation
// Time: O(n), Space: O(n)

class Solution {
    evalRPN(tokens) {
        const stack = [];
        
        for (const token of tokens) {
            if (['+', '-', '*', '/'].includes(token)) {
                const b = stack.pop();
                const a = stack.pop();
                if (token === '+') {
                    stack.push(a + b);
                } else if (token === '-') {
                    stack.push(a - b);
                } else if (token === '*') {
                    stack.push(a * b);
                } else {  // division
                    // Truncate toward zero
                    stack.push(Math.trunc(a / b));
                }
            } else {
                stack.push(parseInt(token));
            }
        }
        
        return stack[0];
    }
}

// Test
if (require.main === module) {
    const sol = new Solution();
    console.log(sol.evalRPN(["2", "1", "+", "3", "*"]));  // 9
    console.log(sol.evalRPN(["4", "13", "5", "/", "+"]));  // 6
}

module.exports = Solution;

File: 13-eval-rpn/solution.cpp
// Evaluate Reverse Polish Notation
// Time: O(n), Space: O(n)

#include <iostream>
#include <vector>
#include <string>
#include <stack>
using namespace std;

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> stk;
        
        for (const string& token : tokens) {
            if (token == "+" || token == "-" || token == "*" || token == "/") {
                int b = stk.top(); stk.pop();
                int a = stk.top(); stk.pop();
                if (token == "+") {
                    stk.push(a + b);
                } else if (token == "-") {
                    stk.push(a - b);
                } else if (token == "*") {
                    stk.push(a * b);
                } else {
                    stk.push(a / b);
                }
            } else {
                stk.push(stoi(token));
            }
        }
        
        return stk.top();
    }
};

int main() {
    Solution sol;
    vector<string> tokens1 = {"2", "1", "+", "3", "*"};
    vector<string> tokens2 = {"4", "13", "5", "/", "+"};
    cout << sol.evalRPN(tokens1) << endl;  // 9
    cout << sol.evalRPN(tokens2) << endl;  // 6
    return 0;
}
